examples/10.dynamic_configuration/config_broker.cc

```diff
@@ -0,0 +1,275 @@
+#include <fail-simulator-on-error.h>
+#include <futex.h>
+#include <thread.h>
+
+#include <string.h>
+#include <vector>
+
+#include "config_broker.h"
+
+// Import some useful things from the CHERI namespace.
+using namespace CHERI;
+
+// Expose debugging features unconditionally for this compartment.
+using Debug = ConditionalDebug<true, "Config Broker ">;
+
+//
+// Data type of config data for a compartment
+//
+struct cbInfo
+{
+	uint16_t id;
+	__cheri_callback void (*cb)(const char *, void *);
+};
+
+struct Config
+{
+	char                        *name;
+	bool                         updated;
+	std::vector<struct cbInfo *> cbList;
+	void                        *data;
+};
+
+//
+// futex to signal when new data is available
+//
+static uint32_t pending = 0;
+
+//
+// Set of config data items.
+//
+std::vector<struct Config *> configData;
+
+//
+// unseal a config capability.
+//
+ConfigToken *config_capability_unseal(SObj sealed_cap)
+{
+	auto            key    = STATIC_SEALING_TYPE(ConfigKey);
+	static uint16_t nextId = 1;
+
+	ConfigToken *token =
+	  token_unseal<ConfigToken>(key, Sealed<ConfigToken>{sealed_cap});
+
+	if (token == nullptr)
+	{
+		Debug::log("invalid config capability {}", sealed_cap);
+		return nullptr;
+	}
+
+	if (token->id == 0)
+	{
+		// Assign an ID so we can track the callbacks added
+		// from this capability
+		token->id = nextId++;
+
+		// Count how many config ids are defined to
+		// save us having to do this each time
+		for (auto i = 0; i < MAX_CONFIG_IDS; i++)
+		{
+			if (strlen(token->configId[i]))
+			{
+				token->count++;
+			}
+		}
+	}
+
+	return token;
+}
+
+//
+// Find a Config by name.  If it doesn't already exist
+// create one.
+//
+Config *find_config(const char *name)
+{
+	for (auto &c : configData)
+	{
+		if (strcmp(c->name, name) == 0)
+		{
+			return c;
+		}
+	}
+
+	// Allocate a Config object
+	Config *c = static_cast<Config *>(malloc(sizeof(Config)));
+
+	// Save the name
+	c->name = static_cast<char *>(malloc(strlen(name)));
+	strncpy(c->name, name, strlen(name));
+
+	c->updated = false;
+	c->data    = nullptr;
+
+	// Add it to the vector
+	configData.push_back(c);
+
+	return c;
+};
+
+//
+// Add a callback to the list for a Config.  Use the
+// id to prevent adding more than one callback for
+// each client
+//
+void add_callback(Config               *c,
+                  uint16_t              id,
+                  __cheri_callback void cb(const char *, void *))
+{
+	for (auto &cb_info : c->cbList)
+	{
+		if (cb_info->id == id)
+		{
+			cb_info->cb = cb;
+			return;
+		}
+	}
+
+	cbInfo *cb_info = static_cast<cbInfo *>(malloc(sizeof(cbInfo)));
+	cb_info->id     = id;
+	cb_info->cb     = cb;
+	c->cbList.push_back(cb_info);
+}
+
+//
+// Set a configuration data item.  The capabailty must have the
+// is_source set to true and include the name of the item being
+// set
+//
+void __cheri_compartment("config_broker")
+  set_config(SObj sealed_cap, const char *name, void *data)
```

Comment:
Note that the sealed objects are mutable once unsealed, so you can put the index into the tokens array into the sealed object and do the O(n) thing only on first use of each token.  We do something similar in the allocator to lazily assign quota IDs.

Reply:
Is there a requirement for a single capability to authorise publishing on multiple topics?  I think a lot of complexity goes away without that.

Reply:
Yeah, will look at moving to a capability per topic.   

I want to keep it so that the broker doesn't need to know what configuration items are defined, such that with this framework in place if I needed to add a new item I don't have to modify the broker code, limiting changes just to the affected producer(s) and consumer(s) - and I think I've worked out how to do that in the macros now ;-)